module/VuFind/src/VuFind/Auth/Email.php

```diff
@@ -0,0 +1,162 @@
+     */
+    public function authenticate($request)
+    {
+        // This is a dual-mode method:
+        // First, try to find a user account with the provided email address and send
+        // a login link.
+        // Second, log the user in with the hash from the login link.
+
+        $email = trim($request->getPost()->get('username'));
+        $hash = $request->getQuery('hash');
+        if (!$email && !$hash) {
+            throw new AuthException('authentication_error_blank');
+        }
+
+        if (!$hash) {
+            // Validate the credentials:
+            $user = $this->getUserTable()->getByEmail($email, false);
+            if ($user) {
+                $loginData = [
+                    'vufind_id' => $user['id']
+                ];
+                $this->emailAuthenticator
+                    ->sendAuthenticationLink($user['email'], $loginData);
+            }
+            // Don't reveal the result
+            throw new \VuFind\Exception\AuthInProgress('email_login_link_sent');
+        }
+
+        $loginData = $this->emailAuthenticator->authenticate($hash);
+        if (isset($loginData['vufind_id'])) {
+            return $this->getUserTable()->getById($loginData['vufind_id']);
+        } else {
+            return $this->processUser($loginData);
+        }
+
+        // If we got this far, we have a problem:
+        throw new AuthException('authentication_error_invalid');
+    }
+
+    /**
+     * Whether this authentication method needs CSRF checking for the request.
+     *
+     * @param \Zend\Http\PhpEnvironment\Request $request Request object.
+     *
+     * @throws AuthException
+     * @return void
+     *
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function needsCsrfCheck($request)
+    {
+        // Disable CSRF if we get a hash in the request
+        return $request->getQuery('hash') ? false : true;
+    }
+
+    /**
+     * Update the database using login user details, then return the User object.
+     *
+     * @param array $info User details returned by the login initiator like ILS.
+     *
+     * @throws AuthException
+     * @return \VuFind\Db\Row\User Processed User object.
+     */
+    protected function processUser($info)
+    {
+        // Check to see if we already have an account for this user:
+        $userTable = $this->getUserTable();
+        if (!empty($info['id'])) {
+            $user = $userTable->getByCatalogId($info['id']);
+            if (empty($user)) {
+                $user = $userTable->getByUsername($info['email']);
+                $user->saveCatalogId($info['id']);
+            }
+        } else {
+            $user = $userTable->getByUsername($info['email']);
+        }
+
+        // No need to store a password in VuFind's main password field:
+        $user->password = '';
+
+        // Update user information based on received data:
+        $fields = ['firstname', 'lastname', 'email', 'major', 'college'];
```

Comment:
Out of curiosity, is there really a scenario where email authentication would cause these values to change? If so, where are the new values coming from?

Reply:
They come from the he ILS. When ILS method is used for email-based authentication, it calls EmailAuthenticator with the patron information from the ILS. Then when the authentication link is clicked, EmailAuthenticator fetches the information from the session and returns it for processing.

Reply:
Okay, that makes sense. Thanks for the clarification!